[FLINK-10827][tests] Add test for duplicate() to SerializerTestBase

StefanRRichter opened the pull request:
## What is the purpose of the change

This PR adds a test for proper implementation of `TypeSerializer#duplicate` to `SerializerTestBase`. The test is based on checking that concurrent use of duplicated type serializers yields correct results and does not throw exceptions.

zentol commented:
Related test-failure:
```
testDuplicate(org.apache.flink.api.java.typeutils.runtime.PojoSubclassSerializerTest)  Time elapsed: 0.101 sec  <<< ERROR!
java.lang.IllegalStateException: Concurrent access to KryoSerializer. Thread 1: Thread-240 , Thread 2: Thread-241
	at org.apache.flink.api.java.typeutils.runtime.kryo.KryoSerializer.enterExclusiveThread(KryoSerializer.java:623)
	at org.apache.flink.api.java.typeutils.runtime.kryo.KryoSerializer.copy(KryoSerializer.java:211)
	at org.apache.flink.api.java.typeutils.runtime.PojoSerializer.copy(PojoSerializer.java:256)
	at org.apache.flink.api.common.typeutils.SerializerTestBase$SerializerRunner.run(SerializerTestBase.java:597)
```

StefanRRichter commented:
Thanks, the failure is great because we probably caught another offender.

StefanRRichter commented:
@zentol Can you take another look? I included the fix for the serializer in as a seperate issue/commit to this PR.

zentol commented on flink-core/src/main/java/org/apache/flink/api/java/typeutils/runtime/PojoSerializer.java:
```diff
@@ -181,11 +181,13 @@ public boolean isImmutableType() {
+			duplicateFieldSerializers = fieldSerializers;
 		}
+
+		// we must create a new instance, otherwise the subclassSerializerCache can create concurrency problems
```
Nope, I don't understand this change :/

The stacktrace shows that the KryoSerializer is used by multiple threads concurrently. `KryoSerializer#duplicate` always returns a new instance, so this method should've already returned a new instance since `stateful` is true.

StefanRRichter replied:
The point is that the Kryo serializer is not coming at all from the `fieldSerializers` that we duplicate. It comes from `subclassSerializerCache` and nothing is duplicated from there, which it also should not.

zentol commented on flink-core/src/test/java/org/apache/flink/api/common/typeutils/SerializerTestBase.java:
```diff
@@ -526,6 +561,55 @@ public void write(DataInputView source, int numBytes) throws IOException {
+						Assert.assertEquals(testItem, copySerdeTestItem);
+					}
+				}
+			} catch (Exception ex) {
```
AssertionErrors (that are thrown by assertEquals) are not exceptions. https://docs.oracle.com/javase/7/docs/api/java/lang/AssertionError.html

StefanRRichter replied:
I feel like we better just transform the assert into a `checkState` that just throws an exception.

StefanRRichter commented on flink-core/src/test/java/org/apache/flink/api/common/typeutils/SerializerTestBase.java:
```diff
@@ -436,6 +441,36 @@ public void testNullability() {
+
+		for (int i = 0; i < numThreads; ++i) {
+			SerializerRunner<T> runner = new SerializerRunner<>(
+				startLatch,
```
What would be the benefit? This feels like it does exactly what we want.

StefanRRichter commented:
@tillrohrmann I addressed the comments, please have another look. I think the duplicate code is correct as it was because `registeredSerializers` etc is handled in the constructor that was use to duplicate.

StefanRRichter commented:
Thanks @tillrohrmann! Merging.
